### PR TITLE
Add preferred name to user import

### PR DESCRIPTION
**Pull request by hortha**
### Context

We’ve added preferred name to the User model but don’t currently have that in our user import process. In addition, we’re also making a change shortly to make the ID name == the DQT name for accounts that have a TRN.

### Changes proposed in this pull request

Amend the user import process to accept an optional PREFERRED_NAME column. When that’s populated, populate the PreferredName property of the created User record.

Amend the validation logic so that FIRST_NAME and LAST_NAME are required when TRN is not-specified and not permitted otherwise. Amend the import process to query the DQT API where a TRN is specified and use the name it returns for the created ID record.

Handle potential rate limiting from calls to the DQT/TRS API.

### Guidance to review

Changes to user import processing and associated tests.

### Checklist

-   [x] Attach to Trello card
-   [ ] Rebased master
-   [ ] Cleaned commit history
-   [x] Tested by running locally
-   [ ] Reminder created to manually clean any removed app settings post deployment

### CODE REVIEW

**Comment by hortha** [comment]
[Add preferred name to user import](https://trello.com/c/H8bGklTR/1107-add-preferred-name-to-user-import)

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+namespace TeacherIdentity.AuthServer.Infrastructure.Http;
+
+public class HttpRateLimitingException : Exception
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs:3]: I don't think we need the `Http`-prefix if it's inside an `Http` namespace. `TooManyRequests` matches the spec for 429 too.
```suggestion
public class TooManyRequestsException : Exception
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public HttpRateLimitingException(TimeSpan? retryAfter, string? message, Exception? innerException)
+        : base(message, innerException)
+    {
+        RetryAfter = retryAfter;
+    }
+
+    private static string DefaultMessage(TimeSpan? retryAfter) =>
+        $"The operation has been rate-limited and should be retried after {retryAfter}";
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs:24]:
```suggestion
        $"The operation has been rate-limited and should be retried{(retryAfter.HasValue ? " after {retryAfter}" : "")}.";
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpResponseMessageExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+namespace TeacherIdentity.AuthServer.Infrastructure.Http;
+
+public static class HttpResponseMessageExtensions
+{
+    public static HttpResponseMessage EnsureSuccessStatusCodeWithRateLimiting(this HttpResponseMessage httpResponseMessage)
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpResponseMessageExtensions.cs:5]:
```suggestion
    public static HttpResponseMessage HandleErrorResponses(this HttpResponseMessage httpResponseMessage)
```

**Review by gunndabad — APPROVED** [review]

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs (hunk 1 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs]: Done

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpResponseMessageExtensions.cs (hunk 3 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpResponseMessageExtensions.cs]: Done

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs (hunk 2 above)
Comment (hortha) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Infrastructure/Http/HttpRateLimitingException.cs]: Done

**Pull request merged**